libraries/chain/include/eosio/chain/backing_store/db_combined.hpp

```diff
@@ -209,76 +194,130 @@ class rocksdb_whole_db_table_collector {
 
 auto process_all = []() { return true; };
 
-enum class key_context { complete, standalone };
+// the context in which the table writer should be reporting to receiver_.add_row(...).  NOTE: for any context indicating
+// that it reports complete/valid table_id_object_views, for it to be valid it must be passed all keys belonging to that
+// table, if not, the count portion of the table_id_object_view will not be accurate
+enum class key_context {
+   complete,           // report keys (via receiver_.add_row) as they are seen, so table will be after its keys (use rocksdb_whole_db_table_collector to reverse this)
+   complete_reverse,   // report keys (via receiver_.add_row) as they are seen, used when reverse iterating through a table space (do not use with rocksdb_whole_db_table_collector)
+   standalone,         // report an incomplete table (only code/scope/table valid) prior to reporting its keys
+   standalone_reverse, // report an incomplete table (only code/scope/table valid) prior to reporting its keys
+   table_only,         // report only the table, primary and secondary keys are only processed enough to report a complete/valid table_id_object_view
+   table_only_reverse  // report only the table, primary and secondary keys are only processed enough to report a complete/valid table_id_object_view
+};
+
+// processes any keys passed to it, reporting them to its receiver as it sees them.  Use a key_context to adjust this behavior,
+// pass in a keep_processing lambda to indicate to one of the walk_*** methods that processing should stop (like for
+// limiting time querying the database for an PRC call)
 template<typename Receiver, typename Function = std::decay_t < decltype(process_all)>>
 class rocksdb_contract_db_table_writer {
 public:
+   template<key_context Context>
+   friend struct writer_impl;
    using key_type = db_key_value_format::key_type;
 
-   rocksdb_contract_db_table_writer(Receiver &r, key_context context = key_context::complete, Function keep_processing = process_all)
-         : receiver_(r), context_(context), keep_processing_(process_all)  {}
+   rocksdb_contract_db_table_writer(Receiver &r, key_context context, Function keep_processing = process_all)
+         : receiver_(r), context_(context), keep_processing_(keep_processing)  {}
+
+   rocksdb_contract_db_table_writer(Receiver &r, Function keep_processing = process_all)
```

Comment:
It can be removed from this ctor, but it is needed for the one that takes the receiver, context and keep_processing. Should probably make sure that this ctor is used at all. If not remove it.

Reply:
It is used in create_deltas.cpp `create_deltas_rocksdb`

Reply:
Removed.